### PR TITLE
[Merged by Bors] - feat: port combinatorics.quiver.single obj

### PR DESCRIPTION
**Pull request by antoinelab01**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by antoinelab01** [comment]
There's one error left about `deriving Unique`. I also needed to add two minor lemmas to be able to fix some proofs, `ext` and `path_to_list_cons`. Should I backport them to mathlib3 or just add a porting note?

**Comment by jcommelin** [comment]
Note that the `start_port.sh` posts a URL that will allow you to open a PR with a title that's already pre-formatted.

File: Mathlib/Combinatorics/Quiver/SingleObj.lean [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+Single object quiver with a given arrows type.
+
+## Main definitions
+
+Given a type `α`, `SingleObj α` is the `unit` type, whose single object is called `star α`, with
+`Quiver` structure such that `star α ⟶ star α` is the type `α`.
+An element `x : α` can be reinterpreted as an element of `star α ⟶ star α` using
+`to_hom`.
+More generally, a list of elements of `a` can be reinterpreted as a path from `star α` to
+itself using `path_equiv_list`.
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Quiver/SingleObj.lean:26]:
```suggestion
itself using `pathEquivList`.
```

File: Mathlib/Combinatorics/Quiver/SingleObj.lean [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+def toPrefunctor : (α → β) ≃ SingleObj α ⥤q SingleObj β
+    where
+  toFun f := ⟨id, f⟩
+  invFun f a := f.map (toHom a)
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+#align quiver.single_obj.to_prefunctor Quiver.SingleObj.toPrefunctor
+
+theorem to_prefunctor_id : toPrefunctor id = 𝟭q (SingleObj α) :=
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Quiver/SingleObj.lean:98]: Please update other names as well.
```suggestion
theorem toPrefunctor_id : toPrefunctor id = 𝟭q (SingleObj α) :=
```

File: Mathlib/Combinatorics/Quiver/SingleObj.lean [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+theorem path_to_list_to_path {x : SingleObj α} (p : Path (star α) x) :
+    listToPath (pathToList p) = p.cast rfl ext :=
+  by
+  induction' p with y z p a ih
+  rfl
+  dsimp at *; rw [ih]
+#align quiver.single_obj.path_to_list_to_path Quiver.SingleObj.path_to_list_to_path
+
+theorem list_to_path_to_list (l : List α) : pathToList (listToPath l) = l :=
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Quiver/SingleObj.lean:143]:
```suggestion
theorem pathToList_listToPath (l : List α) : pathToList (listToPath l) = l :=
```

File: Mathlib/Combinatorics/Quiver/SingleObj.lean [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+/-- Auxiliary definition for `quiver.single_obj.path_equiv_list`.
+Converts a list of elements of type `α` into a path in the quiver `single_obj α`.
+-/
+@[simp]
+def listToPath : List α → Path (star α) (star α)
+  | [] => Path.nil
+  | a :: l => (listToPath l).cons a
+#align quiver.single_obj.list_to_path Quiver.SingleObj.listToPath
+
+theorem path_to_list_to_path {x : SingleObj α} (p : Path (star α) x) :
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Quiver/SingleObj.lean:135]:
```suggestion
theorem listToPath_pathToList {x : SingleObj α} (p : Path (star α) x) :
```

File: Mathlib/Combinatorics/Quiver/SingleObj.lean [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+#align quiver.single_obj.list_to_path_to_list Quiver.SingleObj.list_to_path_to_list
+
+/-- Paths in `single_obj α` quiver correspond to lists of elements of type `α`. -/
+def pathEquivList : Path (star α) (star α) ≃ List α :=
+  ⟨pathToList, listToPath, fun p => path_to_list_to_path p, list_to_path_to_list⟩
+#align quiver.single_obj.path_equiv_list Quiver.SingleObj.pathEquivList
+
+@[simp]
+theorem path_equiv_list_nil : pathEquivList Path.nil = ([] : List α) :=
```
Comment (jcommelin) [inline Mathlib/Combinatorics/Quiver/SingleObj.lean:157]: Please update other names as well
```suggestion
theorem pathEquivList_nil : pathEquivList Path.nil = ([] : List α) :=
```

**Comment by dupuisf** [comment]
Thanks!

bors r+

**Pull request closed without merging**